pkg/controllers/provisioning/scheduling/topology.go

```diff
@@ -18,149 +18,506 @@ import (
-// `hostname` is 0. Thus, we can always improve topology skew (computed against
-// the global minimum) by adding pods to the cluster. We will generate
-// len(pods)/MaxSkew number of domains, to ensure that skew is not violated for
-// new instances.
-func (t *Topology) computeHostnameTopology(topologyGroup *TopologyGroup, constraints *v1alpha5.Constraints) error {
-	domains := []string{}
-	for i := 0; i < int(math.Ceil(float64(len(topologyGroup.Pods))/float64(topologyGroup.Constraint.MaxSkew))); i++ {
-		domains = append(domains, strings.ToLower(randomdata.Alphanumeric(8)))
+// trackInverseAntiAffinity is used to track topologies of inverse anti-affinities. Here the domains & counts track the
+// pods with the anti-affinity.
+func (t *Topology) trackInverseAntiAffinity(ctx context.Context, node *v1.Node, p *v1.Pod) error {
+	// We intentionally don't track inverse anti-affinity preferences.  We're not required to enforce them so it
+	// just adds complexity for very little value.
+	for _, term := range p.Spec.Affinity.PodAntiAffinity.RequiredDuringSchedulingIgnoredDuringExecution {
+		namespaces, err := t.buildNamespaceList(ctx, term.Namespaces, term.NamespaceSelector)
+		if err != nil {
+			return err
+		}
+		namespaces.Insert(p.Namespace)
+
+		topologyGroup, err := t.newForAffinity(term, namespaces, TopologyTypePodAntiAffinity)
+		if err != nil {
+			return err
+		}
+
+		topologyGroup.isInverse = true
+		hash := topologyGroup.Hash()
+		if existing, ok := t.inverseTopologies[hash]; !ok {
+			t.inverseTopologies[hash] = topologyGroup
+			if err := t.initializeTopologyGroup(ctx, topologyGroup); err != nil {
+				return err
+			}
+		} else {
+			topologyGroup = existing
+		}
+		if node != nil {
+			topologyGroup.RecordUsage(node.Labels[topologyGroup.Key])
+		}
+		topologyGroup.AddOwner(client.ObjectKeyFromObject(p))
 	}
-	topologyGroup.Register(domains...)
-	// This is a bit of a hack that allows the constraints to recognize viable hostname topologies
-	constraints.Requirements = constraints.Requirements.Add(v1.NodeSelectorRequirement{Key: topologyGroup.Constraint.TopologyKey, Operator: v1.NodeSelectorOpIn, Values: domains})
 	return nil
 }
 
-// computeZonalTopology for the topology group. Zones include viable zones for
-// the { cloudprovider, provisioner, pod }. If these zones change over time,
-// topology skew calculations will only include the current viable zone
-// selection. For example, if a cloud provider or provisioner changes the viable
-// set of nodes, topology calculations will rebalance the new set of zones.
-func (t *Topology) computeZonalTopology(ctx context.Context, constraints *v1alpha5.Constraints, topologyGroup *TopologyGroup) error {
-	topologyGroup.Register(constraints.Requirements.Zones().UnsortedList()...)
-	if err := t.countMatchingPods(ctx, topologyGroup); err != nil {
-		return fmt.Errorf("getting matching pods, %w", err)
+func (t *Topology) trackPodAffinityTopology(ctx context.Context, p *v1.Pod) error {
```

Comment:
I'm a bit confused why this one doesn't have parity with `trackTopologySpread`. How does `topology` clarify things? Are there pod affinities that we don't model as `topology`? Does it make sense to suffix all of these with `topology`, like `trackTopologySpreadTopology`? 

```suggestion
func (t *Topology) trackPodAffinity(ctx context.Context, p *v1.Pod) error {
```